#Microbiome Analysis
#Adapted from: https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/set-up-and-pre-processing.html


setwd("/users/PAS0107/osu6702/project/microbiome_R/Microbial-bioinformatics-introductory-course-Material-2018/")

# Create Folders as following
#Tables  
dir.create("tables")
# Figures 
dir.create("figures")  
# Phyloseq objects  
dir.create("phyobjects")  
# Custom codes/notes  
dir.create("codes_notes")

library(microbiome) # data analysis and visualisation
library(phyloseq) # also the basis of data object. Data analysis and visualisation
library(microbiomeutilities) # some utility tools 
library(RColorBrewer) # nice color options
library(ggpubr) # publication quality figures, based on ggplot2
library(DT) # interactive tables in html and markdown
library(data.table) # alternative to data.frame
library(dplyr) # data handling  

#2.5 Read input to phyloseq object
#NG-Tax output

ps.ng.tax <- read_phyloseq(otu.file = "./input_data/humanmicrobiome.biom", 
                           taxonomy.file = NULL, 
                           metadata.file = "./input_data/metadata_table.csv", 
                           type = "biom") 

#2.6 Read the tree file.
#Note: Requires a package called ape and the extension has to be “.tre” and not “.tree” (you can just change the name of the file extension)
# Load tree file
library(ape)
## 
## Attaching package: 'ape'
## The following object is masked from 'package:ggpubr':
## 
##     rotate
treefile_p1 <- read.tree("./input_data/humanmicrobiome.tree")
#2.7 Merge into phyloseq object.
ps.ng.tax <-merge_phyloseq(ps.ng.tax,treefile_p1)
# ps.ng.tax is the first phyloseq object.
print(ps.ng.tax)
rank_names(ps.ng.tax) # we check the taxonomic rank information 
datatable(tax_table(ps.ng.tax))  # the table is interactive you can scrol and search thorugh it for details.

#clean the taxonomy table.
tax_table(ps.ng.tax)[,colnames(tax_table(ps.ng.tax))] <- gsub(tax_table(ps.ng.tax)[,colnames(tax_table(ps.ng.tax))],pattern="[a-z]__",replacement="")
tax_table(ps.ng.tax)[tax_table(ps.ng.tax)[,"Phylum"]== "","Phylum"] <- "Unidentified"
datatable(tax_table(ps.ng.tax))  # the table is interactive you can scrol and search thorugh it for details.
# save the pseq object
saveRDS(ps.ng.tax, "./phyobjects/ps.ng.tax.rds")

# check for features of data  
#Notice the Sparsity, it is high for both approaches, 
#the data has many zeros. A common property of amplicon based microbiota data generated by sequencing.
summarize_phyloseq(ps.ng.tax) 


#2.9 Variablity
#We will check the spread of variation for taxa identifed using both approaches. 
#Coefficient of variation (C.V), i.e. sd(x)/mean(x) is a widely used approach to measure heterogeneity in OTU/ASV abundance data. 
#The plot below shows CV-mean(relaive mean abundance) relationship in the scatter plot, where variation is calculated for each OTU/ASV across samples versus mean relative abundance.

#Now plot C.V.
# the plot_taxa_cv will first convert the counts to relative abundances and then calculate the C.V.

p1 <- plot_taxa_cv(ps.ng.tax, plot.type = "scatter")
p1 + scale_x_log10()
## Scale for 'x' is already present. Adding another scale for 'x', which
## will replace the existing scale.


#Moving on to distribution of taxa
# We make a data table with information on the OTUs
ps0_df_taxa <- data.table(tax_table(ps.ng.tax), 
                          ASVabundance = taxa_sums(ps.ng.tax), 
                          ASV= taxa_names(ps.ng.tax))

ps0_tax_plot <- ggplot(ps0_df_taxa, aes(ASVabundance)) + 
  geom_histogram() + ggtitle("Histogram of ASVs (unique sequence) counts (NG-Tax)") + 
  theme_bw() + scale_x_log10() + ylab("Frequency of ASVs") + xlab("Abundance (raw counts)")

print(ps0_tax_plot)
#The ASVs distribution is close to normal distribution compared to OTU distribution for 97% otu-picking approach. 
#This has implications in downstream statistical analysis like differential abundance testing 

#We can also check for prevalance abundance distribtuion of ASVs and OTUs.
p <- plot_taxa_prevalence(ps.ng.tax, "Phylum")
p

#3. ALPHA DIVERSITIES:
#used to identify within individual taxa richness and evenness. 
ps1 <- readRDS("./phyobjects/ps.ng.tax.rds")
# use print option to see the data saved as phyloseq object.
print(ps1)

summary(sample_sums(ps1))
#As is evident there is a large difference in the number of reads. Minimum is 2458 and maximum is 115023!! There is a ~47X difference!
#We can plot the rarefaction curve for the observed ASVs in the entire data set. This is a way to check how has the richness captured in the sequencing effort.
otu_tab <- t(abundances(ps1))
p <- vegan::rarecurve(otu_tab, 
                      step = 50, label = FALSE, 
                      sample = min(rowSums(otu_tab), 
                                   col = "blue", cex = 0.6))

#Since we are comparing different body sites, some are expected to have low bacterial load.
#We will normalize to the lowest depth of at least 2458 reads to keep maximum samples for our anlaysis. 
#This can be varied to remove samples with lower sequencing depth. This decision will depend on the research question being addressed.

#3.1 Equal sample sums
set.seed(9242)  # This will help in reproducing the filtering and nomalisation. 

ps0.rar <- rarefy_even_depth(ps1, sample.size = 2458) ##Please note that the authors of phyloseq do not advocate using this as a normalization procedure, despite its recent popularity.
#More infor on normalization, type: ?rarefy_even_depth()

## You set `rngseed` to FALSE. Make sure you've set & recorded
##  the random seed of your session for reproducibility.
## See `?set.seed`
## ...
## 63OTUs were removed because they are no longer 
## present in any sample after random subsampling
## ...
saveRDS(ps0.rar, "./phyobjects/ps0.rar.rds")
#Check how much data you have now
ps0.rar <- readRDS("./phyobjects/ps0.rar.rds")
print(ps0.rar)

#Quick check taxa prevalence.
p.rar <- plot_taxa_prevalence(ps0.rar, "Phylum")
p.rar

#3.2 Diversities
#3.2.1 Non-phylogenetic diversities
#For more diversity indices please refer to Microbiome Package
#Let us calculate diversity.
hmp.div <- diversities(ps0.rar, index = "all")
datatable(hmp.div)


#####This is one way to plot the data.

# get the metadata out as seprate object
hmp.meta <- meta(ps0.rar)
# Add the rownames as a new colum for easy integration later.
hmp.meta$sam_name <- rownames(hmp.meta)
# Add the rownames to diversity table
hmp.div$sam_name <- rownames(hmp.div)
# merge these two data frames into one
div.df <- merge(hmp.div,hmp.meta, by = "sam_name")
# check the tables
colnames(div.df)

# Now use this data frame to plot 
p <- ggboxplot(div.df, x = "scientific_name", y = "shannon",
               fill = "scientific_name", palette = "jco")
p + rotate_x_text()

colnames(hmp.div)

########Alternative way
# convert phyloseq object into a long data format.  
div.df2 <- div.df[,c("scientific_name", "inverse_simpson", "gini_simpson", "shannon", "fisher", "coverage")]
# the names are not pretty. we can replace them 
colnames(div.df2) <- c("Location", "Inverse Simpson", "Gini-Simpson", "Shannon", "Fisher", "Coverage")
# check
colnames(div.df2)
div_df_melt <- reshape2::melt(div.df2)
head(div_df_melt)

#The diversity indices are stored under column named variable.
# Now use this data frame to plot 
p <- ggboxplot(div_df_melt, x = "Location", y = "value",
               fill = "Location", 
               palette = "jco", 
               legend= "right",
               facet.by = "variable", 
               scales = "free")

p <- p + rotate_x_text() 
# we will remove the x axis lables
p <- p + rremove("x.text")
p

ggsave("./figures/Diversities.pdf", height = 4, width = 10)
lev <- levels(div_df_melt$Location) # get the variables

# make a pairwise list that we want to compare.
L.pairs <- combn(seq_along(lev), 2, simplify = FALSE, FUN = function(i)lev[i])

pval <- list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 0.1, 1), symbols = c("****", "***", "**", "*", "n.s")) 

p2 <- p + stat_compare_means(comparisons = L.pairs, 
                             label = "p.signif", 
                             symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 0.1, 1), 
                                                symbols = c("****", "***", "**", "*", "n.s")))

print(p2)

########
#######3.2.2 Phylogenetic diversity (PD)
##Phylogenetic diversity is calculated using the picante package.
library(picante)
ps0.rar.asvtab <- as.data.frame(ps0.rar@otu_table)
ps0.rar.tree <- ps0.rar@phy_tree
# hmp.meta from previous code chunks

# We first need to check if the tree is rooted or not 
ps0.rar@phy_tree

# it is a rooted tree
df.pd <- pd(t(ps0.rar.asvtab), ps0.rar.tree,include.root=T) # t(ou_table) transposes the table for use in picante and the tre file comes from the first code chunck we used to read tree file (see making a phyloseq object section).
datatable(df.pd)

#now we need to plot PD. Check above how to get the metadata file from a phyloseq object.
# now we need to plot PD
# We will add the results of PD to this file and then plot.
hmp.meta$Phylogenetic_Diversity <- df.pd$PD

#Plot
pd.plot <- ggboxplot(hmp.meta, x = "scientific_name", 
                     y = "Phylogenetic_Diversity",
                     fill = "scientific_name", 
                     palette = "jco",
                     ylab = "Phylogenetic Diversity", 
                     xlab = "Body site",
                     legend = "right")
pd.plot <- pd.plot + rotate_x_text()

pd.plot + stat_compare_means(comparisons = L.pairs, 
                             label = "p.signif", 
                             symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 0.1, 1), 
                                                symbols = c("****", "***", "**", "*", "n.s")))
##NOTE:
#It is always good to check if there is a correlation between increasing library sizes and richness. 
#Observed ASVs and Phylogenetic diversity can be affected by library sizes. 
#It is always good to check for this before making a choice.
lib.div <- diversities(ps1, index = "all")
lib.div2 <- richness(ps1)

# let us add nmber of total reads/samples
lib.div$ReadsPerSample <- sample_sums(ps1)
lib.div$Richness <- lib.div2$`0`
colnames(lib.div)

#We will use ggscatter function from ggpubr package to visualze.
p1 <- ggscatter(lib.div, "shannon", "ReadsPerSample") + 
  stat_cor(method = "pearson")

p2 <- ggscatter(lib.div, "inverse_simpson", "ReadsPerSample",
                add = "loess") + 
  stat_cor(method = "pearson")

p3 <- ggscatter(lib.div, "Richness", "ReadsPerSample",
                add = "loess") + 
  stat_cor(method = "pearson", 
           label.x = 100, 
           label.y = 50000)

ggarrange(p1,p2,p3, ncol=2, nrow = 2)

####4 COMPOSITION PLOTS:

#4.1 Barplot counts
ps1.com <- ps1

# if you have dada2/deblur output and sequences as taxa names, then you can change them as follows:
#taxa_names(ps1.com) <- paste0("ASV_", rownames(tax_table(ps1.com)))

# We need to set Palette
taxic <- as.data.frame(ps1.com@tax_table)  # this will help in setting large color options

#colourCount = length(unique(taxic$Family))  #define number of variable colors based on number of Family (change the level accordingly to phylum/class/order)
#getPalette = colorRampPalette(brewer.pal(12, "Paired"))  # change the palette as well as the number of colors will change according to palette.

taxic$OTU <- rownames(taxic)  # Add the OTU ids from OTU table into the taxa table at the end.
colnames(taxic)  # You can see that we now have extra taxonomy levels.


taxmat <- as.matrix(taxic)  # convert it into a matrix.
new.tax <- tax_table(taxmat)  # convert into phyloseq compatible file.
tax_table(ps1.com) <- new.tax  # incroporate into phyloseq Object

# now edit the unclassified taxa
tax_table(ps1.com)[tax_table(ps1.com)[, "Family"] == "", "Family"] <- "Unclassified family"

# it would be nice to have the Taxonomic names in italics.
# for that we set this
guide_italics <- guides(fill = guide_legend(label.theme = element_text(size = 15, 
                                                                       face = "italic", colour = "Black", angle = 0)))

## Now we need to plot at family level, we can do it as follows:
# first remove the phy_tree
ps1.com@phy_tree <- NULL
# Second merge at family level 
ps1.com.fam <- microbiome::aggregate_taxa(ps1.com, "Family", top = 10)

plot.composition.COuntAbun <- plot_composition(ps1.com.fam) + theme(legend.position = "bottom") + 
  scale_fill_brewer("Family", palette = "Paired") + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90)) + 
  ggtitle("Relative abundance") + guide_italics + theme(legend.title = element_text(size=18))

plot.composition.COuntAbun

####4.2 Barplot relative abundance
#Make it relative abundance

# the previous pseq object ps1.com.fam is only counts.

# Use traqnsform function of microbiome to convert it to rel abun.
ps1.com.fam.rel <- microbiome::transform(ps1.com.fam, "compositional")

plot.composition.relAbun <- plot_composition(ps1.com.fam.rel, 
                                             sample.sort = "scientific_name", 
                                             x.label = "env_material") + theme(legend.position = "bottom") + scale_fill_brewer("Family", palette = "Paired") + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90)) + 
  ggtitle("Relative abundance") + guide_italics + theme(legend.title = element_text(size=18))

plot.composition.relAbun


####4.2.1 Barplot customize
data.com <- plot.composition.relAbun$data
colnames(data.com)
## [1] "OTU"       "Sample"    "Abundance" "xlabel"
p.com <- ggplot(data.com, aes(x = Sample, y = Abundance, fill = OTU))
p.com <- p.com + geom_bar(position = "stack", stat = "identity")
p.com <- p.com + scale_x_discrete(labels = data.com$xlabel, breaks = data.com$Sample)
p.com <- p.com + facet_grid(~xlabel, scales = "free") + theme_bw()
p.com <- p.com + scale_fill_brewer("Family", palette = "Paired") 
p.com <- p.com + rremove("x.text") 

ggsave("./figures/Composition plots.pdf", height = 4, width = 6)

### Heatmaps
#These are a good alternative to barplots (if done right).

# base plot
p.heat <- ggplot(data.com, aes(x = Sample, y = OTU)) + geom_tile(aes(fill = Abundance)) 

# Change color
p.heat <- p.heat + scale_fill_distiller("Abundance", palette = "RdYlBu") + theme_bw() 

# Make bacterial names italics
p.heat <- p.heat + theme(axis.text.y = element_text(colour = 'black', 
                                                    size = 10, 
                                                    face = 'italic')) 
# Make seperate samples based on main varaible
p.heat <- p.heat + facet_grid(~xlabel, 
                              scales = "free") + rremove("x.text") 

p.heat <- p.heat + ylab("Family")

#Clean the x-axis
p.heat <- p.heat + theme(axis.title.x=element_blank(),
                         axis.text.x=element_blank(),
                         axis.ticks.x=element_blank()) 

# Clean the facet label box
p.heat <- p.heat + theme(legend.key = element_blank(), 
                         strip.background = element_rect(colour="black", fill="white"))

print(p.heat)
ggsave("./figures/Heatmap.pdf", height = 4, width = 6)

####Extra
#Following is an example of customizing the plot using ggpubr.

# we use count data at family level from the barplot for counts
ps_df <- microbiomeutilities::phy_to_ldf(ps1.com.fam, transform.counts = "compositional")
## An additonal column Sam_rep with sample names is created for reference purpose
colnames(ps_df)
# this data.frame can be used to customize several plots.  
# example boxplot at family level
p.box <- ggstripchart(ps_df, "scientific_name", "Abundance", 
                      facet.by = "Family", color = "body_product",
                      palette = "jco")

p.box + rremove("x.text")


#########5: BETA DIVERSITY METRICS:
#Beta-diversity: Measures for differences between samples from different groups to identify if there are differences ..
#in the overall community composition and structure.
# read non rarefied data
ps1 <- readRDS("./phyobjects/ps.ng.tax.rds")

# read rarefied data
ps0.rar.rds <- readRDS("./phyobjects/ps0.rar.rds")
# use print option to see the data saved as phyloseq object.

##5.1 Phylogenetic beta-diversity metrics
#5.1.1 Unweighted Unifrac
#Unweighted Unifrac is based on presence/absence of different taxa and abundance is not important. However, it is sensitive to the sequencing depth. 

# if we remove OTUs that are detected atleast 10 times in 5% of the samples
ps0.rar.filtered <- core(ps0.rar.rds, detection = 10, prevalence = 0.05)
summarize_phyloseq(ps0.rar.filtered)
# we reduce the sparsity considerably. 

ordu.unwt.uni <- ordinate(ps0.rar.rds, "PCoA", "unifrac", weighted=F)

# check for Eigen values 
# barplot(ordu.unwt.uni$values$Eigenvalues[1:10])
unwt.unifrac <- plot_ordination(ps0.rar.rds, 
                                ordu.unwt.uni, color="scientific_name") 
unwt.unifrac <- unwt.unifrac + ggtitle("Unweighted UniFrac") + geom_point(size = 2)
unwt.unifrac <- unwt.unifrac + theme_classic() + scale_color_brewer("Location", palette = "Set2")
print(unwt.unifrac)

##5.1.2 Weighted Unifrac
#Weighted Unifrac will consider the abundances of different taxa.
ps1.rel <- microbiome::transform(ps1, "compositional")
ordu.wt.uni <- ordinate(ps1.rel , "PCoA", "unifrac", weighted=T)

# check for Eigen values 
# barplot(ordu.unwt.uni$values$Eigenvalues[1:10])
wt.unifrac <- plot_ordination(ps1.rel, 
                              ordu.wt.uni, color="scientific_name") 
wt.unifrac <- wt.unifrac + ggtitle("Weighted UniFrac") + geom_point(size = 2)
wt.unifrac <- wt.unifrac + theme_classic() + scale_color_brewer("Location", palette = "Set2")
print(wt.unifrac)
print(wt.unifrac + stat_ellipse())

###5.2 Population-level Density landscapes
p <- plot_landscape(ps1.rel, 
                    "NMDS", 
                    "bray", 
                    col = "scientific_name") +
  labs(title = paste("NMDS / Bray-Curtis"))   

p <- p + scale_color_brewer(palette = "Dark2")+ scale_fill_gradient(low = "#e0ecf4", high = "#6e016b") 
p


####5.3 PERMANOVA
##PERMOVA further reading: https://onlinelibrary.wiley.com/doi/10.1002/9781118445112.stat07841
library(vegan)
## Loading required package: permute
## Loading required package: lattice
## This is vegan 2.5-2
metadf <- data.frame(sample_data(ps1.rel))

unifrac.dist <- UniFrac(ps1.rel, 
                        weighted = TRUE, 
                        normalized = TRUE,  
                        parallel = FALSE, 
                        fast = TRUE)

permanova <- adonis(unifrac.dist ~ scientific_name, data = metadf)

permanova
#In the output, scientific_name(gut, oral, skin, vaginal) variable(factor) has signficicant p-value =0.001 
#meaning, changs in bandance can be explained by different body source. 

##5.4 Checking the homogeneity condition
#Type ?betadisper in R console for more information.
ps.disper <- betadisper(unifrac.dist, metadf$scientific_name)
permutest(ps.disper, pairwise = TRUE)


